zkevm-circuits/src/evm_circuit/execution/error_oog_memory_copy.rs

```diff
@@ -0,0 +1,384 @@
+            "Gas left is less than gas cost",
+            insufficient_gas.expr(),
+            1.expr(),
+        );
+
+        let common_error_gadget = CommonErrorGadget::construct(
+            cb,
+            opcode.expr(),
+            // EXTCODECOPY has extra 1 call context lookup (tx_id), 1 account access list
+            // read (is_warm), and 1 stack pop (external_address).
+            5.expr() + 3.expr() * is_extcodecopy.expr(),
+        );
+
+        Self {
+            opcode,
+            is_warm,
+            tx_id,
+            external_address,
+            src_offset,
+            dst_memory_addr,
+            memory_expansion,
+            memory_copier_gas,
+            insufficient_gas,
+            is_extcodecopy,
+            common_error_gadget,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        block: &Block<F>,
+        transaction: &Transaction,
+        call: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        let opcode = step.opcode.unwrap();
+        let is_extcodecopy = usize::from(opcode == OpcodeId::EXTCODECOPY);
+
+        log::debug!(
+            "ErrorOutOfGasMemoryCopy: opcode = {}, gas_left = {}, gas_cost = {}",
+            opcode,
+            step.gas_left,
+            step.gas_cost,
+        );
+
+        let (is_warm, external_address) = if is_extcodecopy == 1 {
+            (
+                block.rws[step.rw_indices[1]].tx_access_list_value_pair().0,
+                block.rws[step.rw_indices[2]].stack_value(),
+            )
+        } else {
+            (false, U256::zero())
+        };
+
+        let rw_offset = 3 * is_extcodecopy;
+        let [dst_offset, src_offset, copy_size] = [rw_offset, rw_offset + 1, rw_offset + 2]
+            .map(|idx| block.rws[step.rw_indices[idx]].stack_value());
+
+        self.opcode
+            .assign(region, offset, Value::known(F::from(opcode.as_u64())))?;
+        self.is_warm
+            .assign(region, offset, Value::known(F::from(u64::from(is_warm))))?;
+        self.tx_id
+            .assign(region, offset, Value::known(F::from(transaction.id as u64)))?;
+        self.external_address
+            .assign(region, offset, Some(external_address.to_le_bytes()))?;
+        self.src_offset
+            .assign(region, offset, Some(src_offset.to_le_bytes()))?;
+        let memory_addr = self
+            .dst_memory_addr
+            .assign(region, offset, dst_offset, copy_size)?;
+        let (_, memory_expansion_cost) =
+            self.memory_expansion
+                .assign(region, offset, step.memory_word_size(), [memory_addr])?;
+        self.memory_copier_gas
+            .assign(region, offset, copy_size.as_u64(), memory_expansion_cost)?;
+        self.insufficient_gas.assign_value(
+            region,
+            offset,
+            Value::known(F::from(step.gas_left)),
+            Value::known(F::from(step.gas_cost)),
```

Comment:
wrong.    when constant_gas_cost < step.gas_left, step.gas_cost will be constant_gas_cost instead of constant_gas_cost + memory_copier_gas...   So you should write like this:

```
let memory_copier_gas = self.memory_copier_gas
            .assign(region, offset, copy_size.as_u64(), memory_expansion_cost)?;
        let constant_gas_cost = if is_extcodecopy {
            if is_warm {
                GasCost::WARM_ACCESS
            } else {
                GasCost::COLD_ACCOUNT_ACCESS
            }
        } else {
            GasCost::FASTEST
        };
        log::trace!("memory_copier_gas {:?} constant_gas_cost {:?}", memory_copier_gas, constant_gas_cost.0);
        self.insufficient_gas.assign_value(
            region,
            offset,
            Value::known(F::from(step.gas_left)),
            Value::known(F::from(memory_copier_gas + constant_gas_cost.0)),
        )?;
```

Reply:
Fixed in commit [9410eda](https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1288/commits/9410eda35dd11f9f8c386aa562dd2c996b461638).

Will also fix previous ones in this PR.

Reply:
@silathdiir  have you tested if memory_copier_gas + constant_gas_cost.0 == step.gas ?

Reply:
> @silathdiir have you tested if memory_copier_gas + constant_gas_cost.0 == step. gas_cost ?

Yes. They are equal. But if the `step.gas_left < const_gas_cost`, `step.gas_cost = const_gas_cost`. Reference the go-ethereum code as:

Reply:
seems to me they don't come to OOG constant gadget even gas left < constant gas according to error  https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/main/bus-mapping/src/error.rs#L137

Reply:
Link to another issue https://github.com/privacy-scaling-explorations/zkevm-circuits/issues/1305